builders/build/substrate/adding-custom-made-module.md

```diff
@@ -0,0 +1,338 @@
+
+1. `#[frame_support::pallet]` - this attribute is the entry point that marks the module as usable in the runtime
+2. `#[pallet::pallet]` - applied to a structure that is used to retrieve module information easily
+3. `#[pallet::config]` - is a required attribute to define the configuration for the data types of the module
+4. `#[pallet::call]` -  this macro is used to define functions that will be exposed as transactions, allowing them to be dispatched to the runtime. It is here that the developers add their custom transactions and logic
+5. `#[pallet::error]` - as transactions may not be successful (insufficient funds, as an error example) and for security reasons, a custom module can never end up throwing an exception, all the possible errors are to be identified and listed in an enum to be returned upon an unsuccessful execution
+6. `#[pallet::event]` - events can be defined and used as a means to provide more information to the user
+7. `#[pallet::storage]` - this macro is used to define elements that will be persisted in storage. As resources are scarce in a blockchain, it should be used wisely to store only sensible information
+
+### Implementing the Module Basic Structure {: #implementing-basic-structure }
+
+The first two mandatory macros, `#[frame_support::pallet]` and `#[pallet::pallet]`, provide the basic structure of the module and are required to enable the module to be used in a Substrate runtime.
+
+The following snippet shows the general structure of a custom Substrate module.
+
+```rust
+#[frame_support::pallet(dev_mode)]
+pub mod pallet {
+    ...
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+    
+    // All the logic goes here
+}
+```
+
+### Implementing the Module Configuration {: #implementing-module-configuration }
+
+To make the modules highly adaptable, their configuration is abstract enough to allow them to be adapted to the specific requirements of the use case the runtime implements.
+
+The implementation of the `#[pallet::config]` macro is mandatory and sets the module's dependency on other modules and the types and values specified by the runtime-specific settings. More about module dependency in the [Substrate documentation](https://docs.substrate.io/build/pallet-coupling/){target=_blank}.
+
+
+In this example, the lottery module depends on other modules to manage the currency and the random function to select the winner and also reads and uses the ticket price and the maximum number of participants directly from the runtime settings. The configurations for this module are:
+
+1. Events: the module depends on the runtime's definition of an event to be able to emit them.
+2. Currency: this module needs to transfer funds, hence, it needs the definition of the currency system from the runtime
+3. Randomness: to fairly select the winner of the prize from the list of participants 
+4. Ticket cost: the price to charge the buyers that participate in the lottery
```

Comment:
Substrate magic: in the runtime configuration we have a line "type TicketCost = ConstU128<1000000000000000>;" that defines the concrete type and value, so the pallet matches its abstract type definition with that concrete one defined in the runtime, using BalanceOf<> (instead of u128) to make it possible to arithmetically operate with the currency underlying platform (in this case, balances)